File: modern/src/main/rs/emulate/modern/script/interp/stack/LongStack.java
package rs.emulate.modern.script.interp.stack;

import java.util.Arrays;
import java.util.NoSuchElementException;

/**
 * A simple stack of {@code long}s, implemented specifically to avoid the boxing overhead from generic types.
 *
 * @author Major
 */
public final class LongStack {

	/**
	 * The pointer to the next position to insert an element in the stack.
	 */
	private int pointer;

	/**
	 * The stack of {@code long}s.
	 */
	private long[] stack;

	/**
	 * Creates the stack with an initial capacity of {@link StackConstants#DEFAULT_SIZE}.
	 */
	public LongStack() {
		this(StackConstants.DEFAULT_SIZE);
	}

	/**
	 * Creates the LongStack with the specified initial capacity.
	 *
	 * @param capacity The initial capacity of the stack.
	 */
	public LongStack(int capacity) {
		stack = new long[capacity];
	}

	/**
	 * Returns whether or not the specified LongStack is equal to this LongStack.
	 * 
	 * @param other The LongStack to compare with.
	 * @return {@code true} if the specified LongStack is equal to this, {@code false} if not.
	 */
	public boolean equals(LongStack other) {
		if (pointer != other.pointer || size() != other.size()) {
			return false;
		}

		return Arrays.equals(stack, other.stack);
	}

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof LongStack) {
			return equals((LongStack) obj);
		}

		return false;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		return (int) Arrays.stream(stack).reduce(prime * stack.length + pointer, (index, result) -> result * prime + index);
	}

	/**
	 * Pops an element from the stack.
	 * 
	 * @return The popped element.
	 * @throws NoSuchElementException If the stack is empty.
	 */
	public long pop() {
		if (pointer == 0) {
			throw new NoSuchElementException("Cannot pop from an empty stack.");
		}

		return stack[--pointer];
	}

	/**
	 * Pushes a value onto this stack.
	 * 
	 * @param value The value.
	 */
	public void push(long value) {
		if (stack.length == size()) {
			grow();
		}

		stack[pointer++] = value;
	}

	/**
	 * Gets the size of this LongStack.
	 * 
	 * @return The size.
	 */
	public int size() {
		return pointer;
	}

	/**
	 * Grows this array.
	 */
	private void grow() {
		stack = Arrays.copyOf(stack, StackConstants.GROWER.apply(stack.length));
	}

}

File: modern/src/main/kotlin/rs/emulate/modern/script/interp/stack/LongStack.kt
package rs.emulate.modern.script.interp.stack

import java.util.Arrays
import java.util.NoSuchElementException

/**
 * A simple stack of `long`s, implemented specifically to avoid the boxing overhead from generic types.
 *
 * @param capacity The initial capacity of the stack.
 */
class LongStack(capacity: Int = StackConstants.DEFAULT_SIZE) {

    /**
     * The pointer to the next position to insert an element in the stack.
     */
    private var pointer: Int = 0

    /**
     * The stack of `long`s.
     */
    private var stack: LongArray = LongArray(capacity)

    /**
     * The size of this LongStack.
     */
    val size: Int
        get() {
            return pointer
        }

    /**
     * Pops an element from the stack.
     */
    fun pop(): Long {
        if (pointer == 0) {
            throw NoSuchElementException("Cannot pop from an empty stack.")
        }

        return stack[--pointer]
    }

    /**
     * Pushes a value onto this stack.
     */
    fun push(value: Long) {
        if (stack.size == size) {
            grow()
        }

        stack[pointer++] = value
    }

    override fun equals(other: Any?): Boolean {
        if (other is LongStack) {
            return pointer == other.pointer && Arrays.equals(stack, other.stack)
        }

        return false
    }

    override fun hashCode(): Int {
        return 31 * pointer + Arrays.hashCode(stack)
    }

    override fun toString(): String {
        return "LongStack(pointer=$pointer, stack=${Arrays.toString(stack)})"
    }

    private fun grow() {
        stack = Arrays.copyOf(stack, StackConstants.GROWER(stack.size))
    }

}
